src/main/java/net/minecraftforge/common/capabilities/CapabilityDispatcher.java

```diff
@@ -5,139 +5,177 @@
+        Map<CapabilityType<?>, IAttachedCapabilityProvider<?, T>> caps = new HashMap<>(other.caps.size(), 1);
+        Map<ResourceLocation, IAttachedCapabilityProvider<?, T>> byName = new HashMap<>(other.byName.size(), 1);
+        for(Map.Entry<ResourceLocation, IAttachedCapabilityProvider<?, T>> entry : other.byName.entrySet())
         {
-            lstCaps.add(parent);
-            if (parent instanceof INBTSerializable)
+            IAttachedCapabilityProvider<?, T> copy = ((IItemStackCapabilityProvider) entry.getValue()).copy(newOwner);
+            // Ideally we would ensure the type and key don't change here, but it's an expensive check that we don't "need" to do.
+            caps.put(copy.getType(), copy);
+            byName.put(copy.getId(), copy);
+        }
+        this.caps = Collections.unmodifiableMap(caps);
+        this.byName = Collections.unmodifiableMap(byName);
+        this.owner = (T) newOwner;
+    }
+
+    /**
+     * Checks if another capability dispatcher is equivalent to this one.<br>
+     * Equivalence in this case means either dispatcher could replace the other one.<br>
+     * This is used in itemstack merging.<br>
+     * 
+     * Equivalence is checked by asking all attached providers to compare against their
+     * presence on the other dispatcher.  Equivalence must be unanimous.
+     * 
+     * @param other The other dispatcher being checked against.
+     * @return If this dispatcher is equivalent to the other dispatcher.
+     * 
+     * @see {@link IItemStackCapabilityProvider#isEquivalentTo(IAttachedCapabilityProvider)}
+     */
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public boolean isEquivalentTo(@Nullable CapabilityDispatcher<ItemStack> other)
+    {
+        if(other == null) {
+            for(IItemStackCapabilityProvider<?> prov : (Collection<IItemStackCapabilityProvider>) (Collection) this.byName.values()) 
             {
-                lstWriters.add((INBTSerializable<Tag>)parent);
-                lstNames.add("Parent");
+                if(!prov.isEquivalentTo(null)) return false;    // Would anyone ever want this behavior? / Do mods have caps that are ignored upon stacking?
             }
+            return true;
         }
-
-        for (Map.Entry<ResourceLocation, ICapabilityProvider> entry : list.entrySet())
+        else
         {
-            ICapabilityProvider prov = entry.getValue();
-            lstCaps.add(prov);
-            if (prov instanceof INBTSerializable)
+            for(ResourceLocation id : this.byName.keySet())
             {
-                lstWriters.add((INBTSerializable<Tag>)prov);
-                lstNames.add(entry.getKey().toString());
+                var ourProv = (IItemStackCapabilityProvider) this.byName.get(id);
+                var theirProv = other == null ? null : (IItemStackCapabilityProvider) other.byName.get(id);        
+                if(ourProv == null && !theirProv.isEquivalentTo(null)) return false;
```

Comment:
How can `ourProv` be null if it's retrieved from `this.byName`?

Reply:
Ah, yeah, that's a remnant from the old method, will drop the dead if-statement